src/main/java/com/mraof/minestuck/item/StructureScannerItem.java

```diff
@@ -1,85 +1,183 @@
-		boolean foundItem = false;
+		if(pPlayer.isCreative() || stack.getDamageValue() != stack.getMaxDamage() && checkFuelNeeded(pPlayer, pLevel))
+		{
+			stack.getOrCreateTag().putBoolean("Powered", true);
+			pLevel.playSound(pPlayer, pPlayer.getX(), pPlayer.getY(), pPlayer.getZ(), SoundEvents.UI_BUTTON_CLICK.get(), SoundSource.AMBIENT, 0.8F, 1.3F);
+			
+			MutableComponent message = Component.translatable("message.temple_scanner.on");
+			pPlayer.sendSystemMessage(message.withStyle(ChatFormatting.DARK_GREEN));
+		}
+		return InteractionResultHolder.success(stack);
 		
-		if(fuelItem != null)
+	}
+	
+	/**
+	 * Check if the item is powered, and if it's out of battery, recharge it.
+	 * Set the location to the nearest structure, check that a structure exists, then reduce charge if fuelled.
+	 */
+	@Override
+	public void inventoryTick(ItemStack pStack, Level pLevel, Entity pEntity, int pSlotId, boolean pIsSelected)
+	{
+		super.inventoryTick(pStack, pLevel, pEntity, pSlotId, pIsSelected);
+		
+		if(pStack.hasTag() && pStack.getTag().getBoolean("Powered") && pLevel instanceof ServerLevel sLevel)
 		{
-			ItemStack fuelStack = new ItemStack(fuelItem.get());
+			resetCharge(pStack);
 			
-			for(ItemStack invItem : playerIn.getInventory().items)
-			{
-				if(ItemStack.isSameItem(invItem, fuelStack))
-				{
-					foundItem = true;
-					if(!levelIn.isClientSide && levelIn.getRandom().nextFloat() >= 0.95F)
-					{
-						invItem.shrink(1);
-						levelIn.playSound(playerIn, playerIn.getX(), playerIn.getY(), playerIn.getZ(), SoundEvents.BLAZE_SHOOT, SoundSource.AMBIENT, 0.4F, 2F);
-					}
-					
-					break;
-				}
-			}
+			BlockPos pos = setLocation(pEntity, sLevel);
+			
+			checkLocation(pStack, pos);
+			
+			reduceCharge(pStack, pEntity, pLevel);
+		}
+	}
+	
+	public static boolean isCharged(ItemStack stack)
+	{
+		return stack.getDamageValue() < stack.getMaxDamage();
+	}
+	
+	public void resetCharge(ItemStack pStack)
+	{
+		if(!isCharged(pStack))
+		{
+			pStack.setDamageValue(0);
+		}
+	}
+	
+	public BlockPos setLocation(Entity pEntity, ServerLevel sLevel)
+	{
+		return sLevel.findNearestMapStructure(structure, pEntity.blockPosition(), 100, false);
+	}
+	
+	public void checkLocation(ItemStack pStack, BlockPos pos)
+	{
+		if(pos == null)
+		{
+			pStack.getTag().remove("TargetLocation");
 		} else
 		{
-			foundItem = true;
+			pStack.getTag().put("TargetLocation", NbtUtils.writeBlockPos(pos));
 		}
-		
-		if(levelIn instanceof ServerLevel level)
+	}
+	
+	/**
+	 * Check that the fuel item is set; if not, don't reduce charge.
+	 * Scanner charge is represented by durability and damage. Damage is dealt every 20 ticks.
+	 * If it runs out of charge, power it off.
+	 *
+	 * @param pStack current item
+	 * @param pEntity player entity
+	 * @param pLevel player's current level
+	 * @return Boolean check for if deactivated, used to reactivate if players still has fuel.
+	 */
+	public boolean reduceCharge(ItemStack pStack, Entity pEntity, Level pLevel)
```

Comment:
do you still plan on using the boolean return? Otherwise this would be suitable to turn into a void

Reply:
Reverted to void.